src/service/presence/mod.rs

```diff
@@ -187,30 +194,31 @@ impl Service {
 		self.db.presence_since(since)
 	}
 
-	async fn handler(&self) -> Result<()> {
-		let mut presence_timers = FuturesUnordered::new();
-		let receiver = self.timer_receiver.lock().await;
-		loop {
-			tokio::select! {
-				event = receiver.recv_async() => {
-
-					match event {
-						Ok((user_id, timeout)) => {
-							debug!("Adding timer {}: {user_id} timeout:{timeout:?}", presence_timers.len());
-							presence_timers.push(presence_timer(user_id, timeout));
-						}
-						Err(e) => {
-							// generally shouldn't happen
-							error!("Failed to receive presence timer through channel: {e}");
-						}
+	#[tracing::instrument(skip_all, name = "presence_loop")]
+	async fn handler_inner(
+		&self, receiver: &MutexGuard<'_, loole::Receiver<(OwnedUserId, Duration)>>,
+		presence_timers: &mut FuturesUnordered<BoxFuture<'_, OwnedUserId>>,
```

Comment:
Not too happy about this, but haven't found a nice way to create a nameable future that returns a value after some time.